### DISCUSSION TITLE
How to render shadow map preview

### DISCUSSION THREAD

**Discussion by TheMostDiligent** [gpuweb/gpuweb#4744] [Q&A] [answered]
Hello.

I need to render a shadow map preview. Normally, I just bind it as a texture and sample it like any other texture with standard bilinear sampler. This does not work in WebGPU:

```
WebGPU: None of the supported sample types (UnfilterableFloat|Depth) of [Texture "Shadow map"] match the expected sample types (Float).
 - While validating entries[1] as a Sampled Texture.
Expected entry layout: {sampleType: TextureSampleType::Float, viewDimension: 2, multisampled: 0}
```

What combination of depth/unfilterable float/filterable float and filtering/non-filtering sampler should I use to make this work?

**Comment by Kangz** [comment]
Various graphics APIs don't allow filtering depth textures without some special builtin calls. In WebGPU to filter depth textures you must bind it as a `texture_depth_2d`, and not a `texture_2d<f32>`.

Various WebGPU samples use shadow maps so you can take inspiration from them as well. https://webgpu.github.io/webgpu-samples/

**Reply by TheMostDiligent** [reply]
What I am trying to do is not to render shadow, but to render shadow map preview on the screen, like this:

![image](https://github.com/gpuweb/gpuweb/assets/61806567/1c233d97-9162-4387-8a2b-86bf44127967)

**Reply by greggman** [reply]
[Depth textures are "unfilterable-float"](https://www.w3.org/TR/webgpu/#depth-formats) (same as "rgba32float" by default). So, they can't be used with a pipeline who's bindGroups expect filtering to work.

To make a pipeline that doesn't expect filtering you have to manually create a bindGroupLayout so that you can mark the texture bindings `samplerType` as `"unfilterable-float"` and the sampler bindings `type` as `"non-filtering"` ('auto' created layouts set these to "float" and "filtering" respectively)

It's the same issue as this thread, your other Q&A : https://github.com/gpuweb/gpuweb/discussions/4726#discussioncomment-9928668

**Reply by greggman** [reply, marked as answer]
Here's a sample: https://jsgist.org/?src=19d21672ec057eb688a9a43f687deb42

**Reply by greggman** [reply]
Or, as @Kangz mentioned, you can use `texture_depth_2d`, then you don't have to create the bind group layout.

Example: https://jsgist.org/?src=588313f832e643a2d42a9aed484e059d

**Reply by TheMostDiligent** [reply]
@greggman Thank you, this helps!

texture_depth_2d can be sampled with a non-comparison sampler?

**Reply by greggman** [reply]
yes, as it is in the example linked.

**Reply by TheMostDiligent** [reply]
Thanks!
